jbmc/src/java_bytecode/java_trace_validation.cpp

```diff
@@ -0,0 +1,255 @@
+/*******************************************************************\
+
+Module: Java trace validation
+
+Author: Jeannie Moulton
+
+\*******************************************************************/
+
+#include "java_trace_validation.h"
+
+#include <algorithm>
+
+#include <goto-programs/goto_trace.h>
+#include <util/byte_operators.h>
+#include <util/expr.h>
+#include <util/expr_util.h>
+#include <util/simplify_expr.h>
+#include <util/std_expr.h>
+
+bool check_symbol_structure(const exprt &symbol_expr)
+{
+  const auto symbol = expr_try_dynamic_cast<symbol_exprt>(symbol_expr);
+  return symbol && !symbol->get_identifier().empty();
+}
+
+/// \return true iff the expression is a symbol or is an expression whose first
+/// operand can contain a nested symbol
+static bool can_contain_symbol_operand(const exprt &expr)
+{
+  return can_cast_expr<member_exprt>(expr) ||
+         can_cast_expr<index_exprt>(expr) ||
+         can_cast_expr<address_of_exprt>(expr) ||
+         can_cast_expr<typecast_exprt>(expr) ||
+         can_cast_expr<symbol_exprt>(expr) ||
+         expr.get(ID_identifier) == ID_byte_extract_little_endian;
+}
+
+optionalt<symbol_exprt> get_inner_symbol_expr(exprt expr)
+{
+  while(expr.has_operands())
+  {
+    expr = expr.op0();
+    if(!can_contain_symbol_operand(expr))
+      return {};
+  }
+  if(!check_symbol_structure(expr))
+    return {};
+  return *expr_try_dynamic_cast<symbol_exprt>(expr);
+}
+
+bool check_member_structure(const exprt &member_expr)
+{
+  if(!can_cast_expr<member_exprt>(member_expr))
+    return false;
+  if(!member_expr.has_operands())
+    return false;
+  const auto symbol_operand = get_inner_symbol_expr(member_expr);
+  return symbol_operand.has_value();
+}
+
+static void raise_error(
+  const std::string &side,
+  const exprt &expr,
+  const std::string &message = "")
+{
+  throw std::runtime_error(
+    "JBMC Check trace assumption failure on " + side + " expression:\n" +
+    expr.pretty() + (message.empty() ? "" : "\n" + message));
+}
+
+bool valid_lhs_expr_high_level(const exprt &lhs)
+{
+  return can_cast_expr<member_exprt>(lhs) || can_cast_expr<symbol_exprt>(lhs) ||
+         can_cast_expr<index_exprt>(lhs) ||
+         lhs.id() == ID_byte_extract_little_endian;
+}
+
+bool valid_rhs_expr_high_level(const exprt &rhs)
+{
+  return can_cast_expr<struct_exprt>(rhs) || can_cast_expr<array_exprt>(rhs) ||
+         can_cast_expr<constant_exprt>(rhs) ||
+         can_cast_expr<address_of_exprt>(rhs) ||
+         can_cast_expr<symbol_exprt>(rhs) ||
+         can_cast_expr<array_list_exprt>(rhs) ||
+         rhs.id() == ID_byte_extract_little_endian;
+}
+
+bool can_evaluate_to_constant(const exprt &expression)
```

Comment:
This isn't tested by your unit tests - please add something that uses this :minidisc: